src/blog/better-class-features.md

```diff
@@ -0,0 +1,332 @@
+
+```cpp
+// Load the private name symbol for `#a` into r1
+LdaImmutableCurrentContextSlot [2]
+Star0
+
+// Use the DefineKeyedOwnProperty bytecode to store 0 as the value of
+// the property keyed by the private name symbol `#a` in the instance,
+// that is, `#a = 0`.
+LdaZero
+DefineKeyedOwnProperty <this>, r0, [0]
+```
+
+When the initializer is executed for enough times, V8 allocates one [feedback vector slot](https://ponyfoo.com/articles/an-introduction-to-speculative-optimization-in-v8) for each field being initialized. The slot contains the key of the field being added (in the case of the private field, the private name symbol) and a pair of [hidden classes](https://v8.dev/blog/fast-properties) between which the instance has been transitioning as the result of field initialization. In subsequent initializations, the IC uses the feedback to see if the fields are initialized in the same order on instances with the same hidden classes. If the initialization matches the pattern that V8 has seen before (which is usually the case), V8 takes the fast path and performs the initialization with pre-generated code instead of calling into the runtime, thus speeding up the operation. If the initialization does not match a pattern that V8 has seen before, it falls back to a runtime call to deal with the slow cases.
+
+### Optimizing named public class fields
+
+To speed up initialization of named public class fields, we reused the existing `DefineNamedOwnProperty` bytecode which calls into `DefineNamedOwnIC` either in the interpreter or through the code compiled from the `JSDefineNamedOwnProperty` IR opcode. These were intended for object literal initializations and essentially implement a specialized case of the `[[CreateDataProperty]]` semantics which is also what the named public class fields are specified to use.
+
+Now when V8 encounters this class:
+
+```js
+class A {
+  #a = 0;
+  b = this.#a;
+}
+```
+
+It generates the following bytecode for the `b = this.#a` initializer:
+
+```cpp
+// Load the private name symbol for `#a`
+LdaImmutableCurrentContextSlot [2]
+
+// Load the value of the property keyed by `#a` from the instance into r2
+// Note: LdaKeyedProperty is renamed to GetKeyedProperty in the refactoring
+GetKeyedProperty <this>, [2]
+
+// Use the DefineKeyedOwnProperty bytecode to store the property keyed
+// by `#a` as the value of the property keyed by `b`, that is, `b = this.#a;`
+DefineNamedOwnProperty <this>, [0], [4]
+```
+
+The original `DefineNamedOwnIC` machinery could not be simply plugged into the handling of the named public class fields, as it needed a few patches to deal with an edge case of the class fields. Previously it expected the target being initialized to be an object created by V8, which was always true for object literals, but the class fields can be initialized on user-defined objects when the class extends a base class whose constructor overrides the target:
```

Comment:
```suggestion
The original `DefineNamedOwnIC` machinery could not be simply plugged into the handling of the named public class fields, since it was originally intended only for object literal initialization. Previously it expected the target being initialized to be an object created by V8, which was always true for object literals, but the class fields can be initialized on user-defined objects when the class extends a base class whose constructor overrides the target:
```

I don't follow the distinction between "object created by V8" and "user-defined objects". I think you mean an ordinary object with no own properties or something?

Reply:
I meant "an object that has not yet been touched by the user since its creation", that's probably clearer

Reply:
That is indeed clearer, let's go with that.